spatialmath/mesh.go

```diff
@@ -29,10 +40,279 @@ func (m *Mesh) Triangles() []*Triangle {
+		closestPt := ClosestPointTrianglePoint(worldTri, pt)
+		if closestPt.Sub(pt).Norm() <= radius+buffer {
+			return true
+		}
+	}
+	return false
+}
+
+// collidesWithMesh checks if this mesh collides with another mesh
+// TODO: This function is *begging* for GPU acceleration.
+func (m *Mesh) collidesWithMesh(other *Mesh, collisionBufferMM float64) bool {
+	// Transform all triangles to world space once
+	worldTris1 := make([]*Triangle, len(m.triangles))
+	for i, tri := range m.triangles {
+		worldTris1[i] = NewTriangle(
+			Compose(m.pose, NewPoseFromPoint(tri.p0)).Point(),
+			Compose(m.pose, NewPoseFromPoint(tri.p1)).Point(),
+			Compose(m.pose, NewPoseFromPoint(tri.p2)).Point(),
+		)
+	}
+
+	worldTris2 := make([]*Triangle, len(other.triangles))
+	for i, tri := range other.triangles {
+		worldTris2[i] = NewTriangle(
+			Compose(other.pose, NewPoseFromPoint(tri.p0)).Point(),
+			Compose(other.pose, NewPoseFromPoint(tri.p1)).Point(),
+			Compose(other.pose, NewPoseFromPoint(tri.p2)).Point(),
+		)
+	}
+
+	// Check if any triangles from either mesh collide.
+	// If two triangles intersect, then the segment between two vertices of one triangle intersects the other triangle.
+	for _, worldTri1 := range worldTris1 {
+		p1 := worldTri1.Points()
+
+		for _, worldTri2 := range worldTris2 {
+			p2 := worldTri2.Points()
+
+			// Check segments from tri1 against tri2
+			for i := 0; i < 3; i++ {
+				start := p1[i]
+				end := p1[(i+1)%3]
+				bestSegPt, bestTriPt := closestPointsSegmentTriangle(start, end, worldTri2)
+				if bestSegPt.Sub(bestTriPt).Norm() <= collisionBufferMM {
+					return true
+				}
+			}
+
+			// Check segments from tri2 against tri1
+			for i := 0; i < 3; i++ {
+				start := p2[i]
+				end := p2[(i+1)%3]
+				bestSegPt, bestTriPt := closestPointsSegmentTriangle(start, end, worldTri1)
+				if bestSegPt.Sub(bestTriPt).Norm() <= collisionBufferMM {
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+// distanceFromMesh returns the minimum distance between this mesh and another mesh.
+func (m *Mesh) distanceFromMesh(other *Mesh) float64 {
+	// Transform all triangles to world space once
+	worldTris1 := make([]*Triangle, len(m.triangles))
+	for i, tri := range m.triangles {
+		worldTris1[i] = NewTriangle(
+			Compose(m.pose, NewPoseFromPoint(tri.p0)).Point(),
+			Compose(m.pose, NewPoseFromPoint(tri.p1)).Point(),
+			Compose(m.pose, NewPoseFromPoint(tri.p2)).Point(),
+		)
+	}
+
+	worldTris2 := make([]*Triangle, len(other.triangles))
+	for i, tri := range other.triangles {
+		worldTris2[i] = NewTriangle(
+			Compose(other.pose, NewPoseFromPoint(tri.p0)).Point(),
+			Compose(other.pose, NewPoseFromPoint(tri.p1)).Point(),
+			Compose(other.pose, NewPoseFromPoint(tri.p2)).Point(),
+		)
+	}
+
+	minDist := math.Inf(1)
+	for _, worldTri1 := range worldTris1 {
+		p1 := worldTri1.Points()
+
+		for _, worldTri2 := range worldTris2 {
+			p2 := worldTri2.Points()
+
+			// Check segments from tri1 against tri2
+			for i := 0; i < 3; i++ {
```

Comment:
Could write a triangle v triangle helper to cut down on repeated code

Reply:
Similar to above, we are in the realm where the extra per-triangle function call has performance implications.